src/overlays/actors/ovl_Obj_Mine/z_obj_mine.c

```diff
@@ -24,131 +56,1178 @@ ActorInit Obj_Mine_InitVars = {
+    ObjMineWaterChain* waterChain = &this->chain.water;
+    ObjMineWaterLink* waterLink;
+    s32 i;
+    f32 wallCheckRadius = this->actor.home.rot.z * 5.0f;
+    f32 linkY;
+    f32 swayVel;
+
+    waterChain->drag = 0.9f;
+    waterChain->swayMax = 0.003f;
+
+    waterChain->swayPhaseVel = RANDU_BITS(13);
+    waterChain->restoreXZ = -0.0002f;
+    waterChain->restoreY = -0.0002f;
+
+    swayVel = waterChain->swayMax * (LINK_SIZE / 2.0f);
+    linkY = this->actor.home.pos.y;
+
+    for (i = 0, waterLink = waterChain->links; i < linkCount; i++, waterLink++) {
+        linkY += LINK_SIZE;
+        waterLink->basis.x.x = 1.0f;
+        waterLink->basis.y.y = 1.0f;
+        waterLink->basis.z.z = 1.0f;
+        waterLink->pos.x = this->actor.home.pos.x;
+        waterLink->pos.y = linkY;
+        waterLink->pos.z = this->actor.home.pos.z;
+        waterLink->velocity.x = (Rand_ZeroOne() - 0.5f) * swayVel;
+        waterLink->velocity.y = (Rand_ZeroOne() - 0.5f) * swayVel;
+        waterLink->velocity.z = (Rand_ZeroOne() - 0.5f) * swayVel;
+    }
+    waterChain->maxY = linkY;
+
+    // Rest position is slightly below max chain length to give slack.
+    waterChain->restY = waterChain->maxY * 0.97f;
+    linkY -= LINK_SIZE * 1.5f;
+    if (waterChain->restY < linkY) { // This should only happen if linkCount > 50
+        waterChain->restY = linkY;
+    }
+
+    if (wallCheckRadius < 0.0f) {
+        waterChain->wallCheckDistSq = -1.0f;
+    } else if (wallCheckRadius <= WATER_RADIUS + 1.0f) {
+        waterChain->wallCheckDistSq = 0.0f;
+    } else {
+        waterChain->wallCheckDistSq = SQ(wallCheckRadius - WATER_RADIUS);
+    }
+}
+
+void ObjMine_Water_InitCollider(ObjMine* this, s32 linkCount) {
+    Matrix_Translate(this->actor.home.pos.x, this->actor.home.pos.y + (linkCount * LINK_SIZE) + ATTACH_OFFSET,
+                     this->actor.home.pos.z, MTXMODE_NEW);
+    Matrix_Scale(this->actor.scale.x, this->actor.scale.y, this->actor.scale.z, MTXMODE_APPLY);
+    Collider_UpdateSpheres(0, &this->collider);
+}
+
+void ObjMine_Water_SetWorld(ObjMine* this) {
+    s32 pad; // Could be recast to thisx as in ObjMine_Path_Move
+    ObjMineWaterChain* waterChain = &this->chain.water;
+    s32 linkCount = OBJMINE_GET_LINK_COUNT(&this->actor);
+
+    if (linkCount == 0) {
+        this->actor.world.pos.y = this->actor.home.pos.y + LINK_SIZE + ATTACH_OFFSET;
+    } else {
+        ObjMineWaterLink* lastLink = &waterChain->links[linkCount - 1];
+        Vec3f mineOffset;
+
+        Math_Vec3f_ScaleAndStore(&lastLink->basis.y, ATTACH_OFFSET, &mineOffset);
+        Math_Vec3f_Sum(&lastLink->pos, &mineOffset, &this->actor.world.pos);
+    }
+}
+
+void ObjMine_Water_WallCheck(ObjMine* this, PlayState* play) {
+    s32 pad; // Could be recast to thisx as in ObjMine_Path_Move
+    ObjMineWaterChain* waterChain = &this->chain.water;
+
+    waterChain->touchWall = false;
+    if (waterChain->wallCheckDistSq > -1e-6f) {
```

Comment:
ohh, very small float

Reply:
The code is full of these "let's not compare directly to 0" checks, though the small constant used is inconsistent. This one is kind of amusing because it's implementing >= 0. Apparently the dev thought that the = in >= was enough to not use direct 0 comparison.

Reply:
```suggestion
    if (waterChain->wallCheckDistSq > -SQ(0.001)) {
```
And other checks against `wallCheckDistSq`

Reply:
Tried that. It doesn't match as `SQ(0.001f)`. Not that it matters, as this isn't actually comparing to a squared value--just verifying that it's positive because `-1` is used to indicate no checks are to be made.

Reply:
I think `1e-6f` is used enough in here that it could get a define, either for the positive or negative version.